integration_test/util/testApolloClient.js

```diff
@@ -0,0 +1,62 @@
+const { split } = require('apollo-link')
+const { HttpLink } = require('apollo-link-http')
+const { WebSocketLink } = require('apollo-link-ws')
+const { getMainDefinition } = require('apollo-utilities')
+const fetch = require('node-fetch')
+const ws = require('ws')
+const { ApolloClient } = require('apollo-client')
+const { InMemoryCache } = require('apollo-cache-inmemory')
+
+class TestApolloClient {
+  constructor (host = 'localhost:8000') {
+    this.client = createApolloClient(host)
+  }
+
+  subscribe (query, timeout = 3000) {
+    return new Promise((resolve, reject) => {
+      this.client.subscribe({
+        query: query
+      }).subscribe({
+        next: resolve,
+        error: reject
+      })
+      setTimeout(reject.bind(null, new Error('timed out while waiting for subscription result')), timeout)
+    })
+  }
+}
+
+module.exports = TestApolloClient
+
+function createApolloClient (host) {
+  // Create an http link:
+  const httpLink = new HttpLink({
+    uri: `http://${host}/graphql`,
+    fetch: fetch
+  })
+
+  // Create a WebSocket link:
+  const wsLink = new WebSocketLink({
+    uri: `ws://${host}/subscriptions`,
+    options: {
+      reconnect: true
+    },
+    webSocketImpl: ws
+  })
+
+  // using the ability to split links, you can send data to each link
+  // depending on what kind of operation is being sent
+  const link = split(
+    // split based on operation type
+    ({ query }) => {
+      const { kind, operation } = getMainDefinition(query)
+      return kind === 'OperationDefinition' && operation === 'subscription'
+    },
+    wsLink,
+    httpLink
+  )
+
+  return new ApolloClient({
+    link: link,
+    cache: new InMemoryCache()
```

Comment:
Is that required for tests?

Reply:
This is boilerplate code taken directly from the Apollo docs to set up a client that can talk via http and via websockets. It looks really weird but I guess it's because they have support for using any http lib and any websocket lib. (For compatibility across browsers and node).

I've also created a wrapper class around the Apollo client that 'Promisifies' the subscribe method in order to make testing of subscriptions a bit simpler. See in the tests how I can do something like `const result = await subscription`